net.go

```diff
@@ -78,3 +82,96 @@ func Listen(network, address string) (net.Listener, error) {
 	// I asked about it here: https://groups.google.com/g/golang-nuts/c/Q1I7Viz9AJc
 	return defaultListenConfig().Listen(context.Background(), network, address)
 }
+
+type ListenerCallbacks interface {
+	// OnAccept is called when a new connection is successfully accepted.
+	OnAccept(net.Conn)
+
+	// OnTLSHandshakeError is called after a TLS handshake errors out.
+	OnTLSHandshakeError(*tls.Conn, error)
+}
+
+type Listener struct {
+	Address          string
+	Log              log.Logger
+	TLSConfig        *tls.Config
+	HandshakeTimeout time.Duration
+	ReadLimit        int64
+	WriteLimit       int64
+	Callbacks        ListenerCallbacks
+
+	listener net.Listener
+}
+
+func (l *Listener) Listen() error {
+	ll, err := Listen("tcp", l.Address)
+	if err != nil {
+		return err
+	}
+
+	if rl, wl := l.ReadLimit, l.WriteLimit; rl > 0 || wl > 0 {
+		// Notice that the ReadLimit stands for the read limit *from* a proxy, and the WriteLimit
```

Comment:
Let's invert and document that.

ReadLimit is for reading form the returned connections and is shared for all connections and WriteLimit is for writing to the returned connections.

Reply:
It's not the scope of this PR. Here I only moved the place where we apply it. I can make an issue to invert that.

Reply:
ok